src/ucs/sys/sock.c

```diff
@@ -951,6 +959,66 @@ int ucs_sockaddr_ip_cmp(const struct sockaddr *sa1, const struct sockaddr *sa2)
                   UCS_IPV4_ADDR_LEN : UCS_IPV6_ADDR_LEN);
 }
 
+ucs_status_t ucs_sockaddr_subnet_match(const struct sockaddr *sa1,
+                                       const struct sockaddr *sa2,
+                                       unsigned prefix_len, int *is_match_p)
+{
+    int matched         = 0;
+    ucs_status_t status = UCS_OK;
+    char ip1_str[UCS_SOCKADDR_STRING_LEN];
+    char ip2_str[UCS_SOCKADDR_STRING_LEN];
+    const void *ipaddr1, *ipaddr2;
+    size_t addr_size, addr_size_bits;
+
+    if (!ucs_sockaddr_are_both_known_af(sa1, sa2)) {
+        status = UCS_ERR_INVALID_PARAM;
+        goto out;
+    }
+
+    if (sa1->sa_family != sa2->sa_family) {
+        ucs_debug("different addr_family: s1 %s s2 %s",
+                  ucs_sockaddr_address_family_str(sa1->sa_family),
+                  ucs_sockaddr_address_family_str(sa2->sa_family));
+        status = UCS_ERR_INVALID_PARAM;
+        goto out;
+    }
+
+    /* Get address size */
+    ucs_assert_always(ucs_sockaddr_inet_addr_sizeof(sa1, &addr_size) == UCS_OK);
+    addr_size_bits = addr_size * UCS_SOCKET_BYTE_TO_BITS;
+
+    /* Check subnet mask size (bits belonging to the prefix) */
+    if (prefix_len > addr_size_bits) {
+        ucs_error("prefix is bigger than address size (in bits): prefix=%u"
+                  " size=%lu",
+                  prefix_len, addr_size_bits);
+        status = UCS_ERR_EXCEEDS_LIMIT;
+        goto out;
+    }
+
+    ipaddr1 = ucs_sockaddr_get_inet_addr(sa1);
+    ipaddr2 = ucs_sockaddr_get_inet_addr(sa2);
+    ucs_assert((ipaddr1 != NULL) && (ipaddr2 != NULL));
+
+    /* Check if the addresses have matching prefixes */
+    matched = ucs_bitwise_is_equal(ipaddr1, ipaddr2, prefix_len);
+
+    if ((ucs_sockaddr_get_ipstr(sa1, ip1_str, UCS_SOCKADDR_STRING_LEN) ==
```

Comment:
check for debug enablement before generating ipstr with `ucs_log_is_enabled(log_level)`